src/tiled/stampbrush.cpp

```diff
@@ -385,7 +388,47 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
             const TileStampVariation variation = mStamp.randomVariation();
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
-            TileLayer *stamp = variation.tileLayer();
+            TileLayer *stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+
+            //If using a hexagonal map, this takes into acount staggering
+            if (mapDocument()->map()->orientation() == 4 && stamp->height()>1) {
+                int shiftMode;//-1 do not shift, 0 shift even rows (relatived to the stamp), 1 shift odd rows
+                int rowOfTop = p.y()-stamp->height()/2;
+                if (rowOfTop<0)
+                    rowOfTop*=-1; //Only used to check if even or odd, and being negative can mess this up.
+
+                if (mCaptureTopCorner == -1)
```

Comment:
It's interesting how this breaks the shift adjustment for captures made starting from -1, just outside the map.

In any case, you will need to associate this data with the tile layer instance rather than storing it as a member of the stamp brush. In fact, the tile layer already has a member for exactly that, which is the `staggerIndex`. All you should need to do, is to make sure this property gets set correctly when the stamp is captured, and to account for it when the stamp is used at odd indexes (or at even indexes, when the stamp stagger index != the map's stagger index).